Introduce stricter checks for might_permit_raw_init under a debug flag

5225225 opened the pull request:
This is intended to be a version of the strict checks tried out in #79296, but also checking number validity (under the assumption that `let _ = std::mem::uninitialized::<u32>()` is UB, which seems to be what https://github.com/rust-lang/unsafe-code-guidelines/issues/71 is leaning towards.)

rust-highfive commented:
Some changes occured to rustc_codegen_cranelift

cc @bjorn3

Some changes occured to the CTFE / Miri engine

cc @rust-lang/miri

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6559818495?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e68b692d3103bb0002e50143e06449cbffddfbd6/checks/6559818495/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy check
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 629 error codes
tidy error: /checkout/src/test/ui/intrinsics/panic-uninitialized-zeroed-strict.rs: ignoring line length unnecessarily
Found 504 error codes
Found 0 error(s) in error codes
Done!
* 375 features
```

</details>

5225225 commented:
@rustbot label +S-waiting-on-author

5225225 commented:
@rustbot label -S-waiting-on-review

RalfJung commented on compiler/rustc_target/src/abi/mod.rs:
```diff
@@ -1468,7 +1477,7 @@ impl<'a, Ty> TyAndLayout<'a, Ty> {
     /// create an all-0/all-undef constant and run the const value validator to see if
     /// this is a valid value for the given type.
-    pub fn might_permit_raw_init<C>(self, cx: &C, zero: bool) -> bool
+    pub fn might_permit_raw_init<C>(self, cx: &C, zero: bool, strict: bool) -> bool
```
`strict` is unused except for the recursive call, isn't it?

RalfJung commented on src/test/ui/intrinsics/panic-uninitialized-zeroed-strict.rs:
```diff
@@ -0,0 +1,95 @@
+        test_panic_msg(
+            || mem::zeroed::<[fn(); 1]>(),
+            "attempted to zero-initialize type `[fn(); 1]`, which is invalid"
+        );
```
This errors even without the strict flag, right?

5225225 replied:
It does not, because it's in an array.

We currently just give up for arrays, the type in them is completely unchecked, you can make an uninit `[String; 1]` just fine if you want. (Except for uninhabited types, you *can't* make a `[!; 1]`)

RalfJung replied:
Ah right.

RalfJung commented on src/test/ui/intrinsics/panic-uninitialized-zeroed-strict.rs:
```diff
@@ -0,0 +1,95 @@
+    Right(i64),
+}
+#[allow(dead_code, non_camel_case_types)]
+enum LR_NonZero {
```
This type is unused in this test.

RalfJung commented on src/test/ui/intrinsics/panic-uninitialized-zeroed-strict.rs:
```diff
@@ -0,0 +1,95 @@
+enum OneVariant { Variant(i32) }
+
+#[allow(dead_code, non_camel_case_types)]
+enum OneVariant_NonZero {
```
This type is unused in this test.

RalfJung commented on src/test/ui/intrinsics/panic-uninitialized-zeroed-strict.rs:
```diff
@@ -0,0 +1,95 @@
+// An `Aggregate` abi enum where 0 is not a valid discriminant.
+#[allow(dead_code)]
+#[repr(i32)]
+enum NoNullVariant {
```
This type is unused in this test.

5225225 commented:
I'll leave the variant checks for future work.

Was initially going to try and get them in now (so it's not as much breakage to add them later), but I figure adding more panics on UB when you're using a debugging flag specifically to panic on UB is pretty justifiable.

That, and I want to get the infrastructure in here sooner than later. Will clean up this PR to just do the array and primitive checks.

5225225 commented:
@rustbot ready

oli-obk commented on compiler/rustc_const_eval/src/interpret/intrinsics.rs:
```diff
@@ -418,7 +422,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                    && !layout.might_permit_raw_init(
+                        self,
+                        /*zero:*/ false,
+                        self.tcx.sess.opts.debugging_opts.strict_init_checks,
```
somewhat annoying to pass this everywhere, when `self` is also passed. Though `HasDataLayout` doesn't have a useful way to get at this.

maybe add a field to [`TargetDataLayout`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/struct.TargetDataLayout.html#) for that bool?

Idk, doesn't seem great either. I guess let's roll with the PR as it is.

bors commented:
:pushpin: Commit d59cc92d7a30f0b4b7d9e192ca0e165a306f61ef has been approved by `oli-obk`

<!-- @bors r=oli-obk d59cc92d7a30f0b4b7d9e192ca0e165a306f61ef -->
<!-- homu: {"type":"Approved","sha":"d59cc92d7a30f0b4b7d9e192ca0e165a306f61ef","approver":"oli-obk"} -->

oli-obk commented:
> [davidtwco](https://github.com/davidtwco) approved these changes now

oh lol, apologies, that's funny timing

RalfJung commented:
Cc https://github.com/rust-lang/rust/issues/66151

bjorn3 commented on compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs:
```diff
@@ -664,7 +664,7 @@ fn codegen_regular_intrinsic_call<'tcx>(
             }
 
-            if intrinsic == sym::assert_zero_valid && !layout.might_permit_raw_init(fx, /*zero:*/ true) {
+            if intrinsic == sym::assert_zero_valid && !layout.might_permit_raw_init(fx, /*zero:*/ true, fx.tcx.sess.opts.debugging_opts.strict_init_checks) {
```
Would you mind formatting this such that the line length stays below 100 characters? This is inside a macro, so rustfmt doesn't pick up on it. Alternatively I can do this myself after this PR gets merged. Not a big deal.

RalfJung replied:
Btw, does cranelift not use the codegen_ssa crate? I was surprised to see the logic duplicated here.

5225225 replied:
Ah, my bad, didn't spot that, sorry!

Bors was already kicked off so I don't think I can change it now, but if bors comes back with an error I need to fix I'll format this code at the same time. Otherwise, best for you to do it.

5225225 replied:
Also, @bjorn3, I was planning to change the `zero` bool parameter here to a `enum ValidAs { Zero, Uninit }` for clarity, seeing as we need to comment what it actually is at the moment, but that requires me to add a `use rustc_target::abi::ValidAs` to the rustc_codegen_cranelift code somewhere.

I see there's an `prelude` module, should I be adding it there or did you want me to just add it in the one file it's used (`intrinsics/mod.rs`)

bjorn3 replied:
> Btw, does cranelift not use the codegen_ssa crate? I was surprised to see the logic duplicated here.

Only the linker and metadata code are shared at the moment. cg_ssa is too much tied to LLVM at the moment. I did some refactorings for this quite a while ago, but didn't really do much about it recently. cg_clif also has a couple of peephole optimizations during codegen that cg_ssa is missing. And finally cg_clif directly generates ssa form where possible, while cg_ssa depends on a mem2reg pass that cranelift is missing.

bjorn3 replied:
> I've just put it in the top of the file where it's needed for now.

Makes sense. The prelude is just for things used in many modules. I should probably make it smaller.

RalfJung commented:
@bors r-

We're not in a rush for this feature I think, so let's fix that nit before landing. :)

RalfJung commented:
And also thanks a lot @5225225 for doing this, I quite like the idea of opting-in to the more strict checks. :D 

FWIW, as @oli-obk proposed once, there is an elegant way to get the "fully strict" check but it requires more work and no code sharing with the non-strict case: fire up an `InterpCx` and create an actual zeroed/uninit allocation, and then ask the validity checker in the interpreter whether this is valid for the given type. That will handle *all* types properly; it is the same logic that Miri uses to detect incorrect values.
If that peeks your interest, feel free to ask about details on Zulip. :D

RalfJung commented on compiler/rustc_target/src/abi/mod.rs:
```diff
@@ -894,6 +894,15 @@ impl Scalar {
+
+    /// Returns `true` if this type can be left uninit.
+    #[inline]
+    pub fn is_uninit_valid<C: HasDataLayout>(&self, _cx: &C) -> bool {
```
Why does this even take the `_cx` parameter?

5225225 commented:
Okay, I've fixed the formatting in the cranelift code, moved the `zero: bool` to a `valid_as: ValidAs` enum (hope you don't mind, but I was touching this code anyways :) ), and removed the redundant parameter on `is_uninit_valid`. 

And yeah, the "actually make an allocation and then ask the interpreter if it's valid" seems like the cleanest solution.

RalfJung commented on compiler/rustc_target/src/abi/mod.rs:
```diff
@@ -1355,6 +1364,14 @@ pub struct PointeeInfo {
+/// Used in `might_permit_raw_init` to indicate the kind of initialisation
+/// that is checked to be valid
+#[derive(Copy, Clone, Debug)]
+pub enum ValidAs {
```
:+1: for making this an enum. However, I don't quite understand `ValidAs` as the name. What about `InitKind`?

5225225 replied:
It was along the lines of "is this type valid as zeroed, or valid as uninit".

In any case, I changed it, I'm not too bothered what it's called.

bjorn3 reviewed (approved):
cg_clif changes LGTM

5225225 commented on compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs:
```diff
@@ -58,6 +58,7 @@ pub(crate) use llvm::codegen_llvm_intrinsic_call;
 use rustc_middle::ty::print::with_no_trimmed_paths;
 use rustc_middle::ty::subst::SubstsRef;
 use rustc_span::symbol::{kw, sym, Symbol};
+use rustc_target::abi::ValidAs;
```
woops forgot to change this

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit dd9f31d000a33c52383f9af9e1dbf44f754590c3 has been approved by `oli-obk`

<!-- @bors r=oli-obk dd9f31d000a33c52383f9af9e1dbf44f754590c3 -->
<!-- homu: {"type":"Approved","sha":"dd9f31d000a33c52383f9af9e1dbf44f754590c3","approver":"oli-obk"} -->
